Add support for asynchronous retrieval from `RedisCache`

jxblum opened the pull request:
See #2650

mp911de commented on src/test/java/org/springframework/data/redis/cache/DefaultRedisCacheWriterUnitTests.java:
```diff
@@ -86,14 +121,236 @@ void getWithNullTtl() {
+		assertThat(value).isNotNull();
+		assertThat(new String(value)).isEqualTo("test");
+
+		verify(this.mockReactiveConnectionFactory, times(1)).getReactiveConnection();
```
I wonder how maintainable this will remain when we have almost 15 verifications and what net value of the test remains. 

Generally looking at the new tests, there's 90% infrastructure and only 10% of actual test code left.

mp911de reviewed (changes requested):
I did a first review pass. We need to come up with a design that doesn't introduce reactive types for all of our users as that will introduce a dependency regression.

jxblum commented on src/main/java/org/springframework/data/redis/cache/BatchStrategies.java:
```diff
@@ -68,6 +65,10 @@ public static BatchStrategy scan(int batchSize) {
 		return new Scan(batchSize);
 	}
 
+	private BatchStrategies() {
```
Understood. Out of curiosity, where is this documented?

mp911de replied:
https://github.com/spring-projects/spring-framework/wiki/Code-Style#java-source-file-organization, as we didn't define anything beyond https://github.com/spring-projects/spring-data-build/blob/main/CONTRIBUTING.adoc#source-code-style

jxblum replied:
I noticed that we are not honoring several of these Code Style guidelines:

* Import order (e.g. Java/Javax/Jakarta library imports first!)
* Location of getter/setters for fields.
* No line break after closing `}` and `else` / `catch` blocks.
* Our Javadoc formatting is not consistent.
* Etc.

jxblum commented on src/main/java/org/springframework/data/redis/cache/DefaultRedisCacheWriter.java:
```diff
@@ -149,6 +137,103 @@ public byte[] get(String name, byte[] key, @Nullable Duration ttl) {
 		return result;
 	}
 
+	@Override
```
I need to think on this further, but my first reaction (non pun intended) is to use delegation or more effective use of `CompletableFuture`.

jxblum commented on src/main/java/org/springframework/data/redis/cache/DefaultRedisCacheWriter.java:
```diff
@@ -149,6 +137,103 @@ public byte[] get(String name, byte[] key, @Nullable Duration ttl) {
+			ByteBuffer wrappedKey = ByteBuffer.wrap(key);
+
+			// Do the same lock check as the regular Cache.get(key); be careful of blocking!
+			Mono<ByteBuffer> cacheLockCheckMono = Mono.fromFuture(CompletableFuture.supplyAsync(() -> {
```
My knowledge of Reactor is still limited in regards to the proper way to do things.

Effectively, I want the `Mono` to wait for the cache lock to be released (if the cache by name was in fact locked in the first place) **before** attempting to do the caching operation (i.e. Redis `get[Ex](..)` command). So, I wouldn't really say it is looping so much as it is waiting, even though inside `checkAndPotentiallyWaitUntilUnlocked(..)` the method loops waiting for the lock to be release. Still, from the caller's perspective, it is just perceived as a "blocking" method.

I wasn't sure how to add an additional operation (i.e. checking the cache lock) to the `Mono` returned from Lettuce before invoking the Redis.get[Ex](..)` command in a asynchronous, non-blocking way. I tried `Mono.doFirst(:Runnable)` but that blocks before subscribing.

I may need some extra guidance here.

mp911de replied:
Probably this needs some refinement, but as a starting point I would suggest:

```
Mono<Boolean> isLocked = …;

isLocked.flatMap(hasLock -> {
	
	if(!hasLock){
		return Mono.empty();
	}

	return Flux.interval(sleepTime).takeUntilOther(isLocked).then();
});
```

jxblum replied:
I think I figured out what I need to do with this. In a nutshell, I used `Flux.interface(delay, period).takeUntil(..)`. The tests passed as I expected them to.

jxblum commented on src/main/java/org/springframework/data/redis/cache/DefaultRedisCacheWriter.java:
```diff
@@ -149,6 +137,103 @@ public byte[] get(String name, byte[] key, @Nullable Duration ttl) {
+	// Execution Strategy applied when Jedis (non-Reactive driver) is used.
+	// Treats API consistently (that is, using Reactive types, such as Mono) at the RedisCacheWriter level
+	// whether "technically Reactive" or not; clearly Jedis is not "Reactive".
+	private BiFunction<byte[], Duration, Mono<byte[]>> asyncExecutionStrategy(String cacheName) {
```
2 things. 

First, my initial reaction was how to support our **Jedis** users, too. So this was the design I came up with.

Second, Spring Framework's contract for the `Cache.retrieve(..)` operation (e.g. [Javadoc](https://docs.spring.io/spring-framework/docs/6.1.0-SNAPSHOT/javadoc-api/org/springframework/cache/Cache.html#retrieve(java.lang.Object))) is very clear... it must not block.

In fact, the approach I took (using `CompletableFuture.supplyAsync(..)` for Jedis) is precisely how Spring Framework handles asynchronous, non-blocking behavior in `Cache.retrieve(..)` for caching providers that do not support asynchronous, non-blocking operations, such as `ConcurrentMapCache` ([source](https://github.com/spring-projects/spring-framework/blob/v6.1.0-M5/spring-context/src/main/java/org/springframework/cache/concurrent/ConcurrentMapCache.java#L159-L171)). In contrast, if the underlying caching provider, such as Caffeine, supports asynchronous, non-blocking operations (in other words, return types, such as `CompletableFuture`), then the implementation is trivial ([source](https://github.com/spring-projects/spring-framework/blob/v6.1.0-M5/spring-context-support/src/main/java/org/springframework/cache/caffeine/CaffeineCache.java#L138-L152)).

jxblum replied:
So, in conclusion, I would say.

Either, we support Jedis users by wrapping the simple cache op in an asynchronous manner, not unlike `ConcurrentMapCache`, or we do not support Jedis at all.

It would be wrong to support Jedis, but only in a blocking manner, which would violate the contract of `Cache.retrieve(..)`.

I am leaning towards supporting Jedis users where possible and the implementation is rather simple.

mp911de replied:
> Either, we support Jedis users by wrapping the simple cache op in an asynchronous manner, not unlike ConcurrentMapCache, or we do not support Jedis at all.

We decided against pretending our blocking infrastructure would support asynchronicity to avoid raising false expectations. `RestTemplate`, JPA, `JdbcTemplate` are several examples that have followed this pattern.

It is not wise to do things just because we can.

> It would be wrong to support Jedis, but only in a blocking manner, which would violate the contract of Cache.retrieve(..).

`retrieve` is allowed to throw `UnsupportedOperationException`. The JCache Cache doesn't support async operations.

jxblum replied:
> _It is not wise to do things just because we can._

No need to state the obvious. Also, I was not part of the original/initial "decision" so I am largely unaware and unfamiliar with what the intended direction is for Spring Data Redis. I am simply acting in the best interest of our community, and I am simply stating my case in favor of this approach, not because we "can".

Also, many caching providers don't support async caching operations yet. Neither did the `ConcurrentMapCache` implementation in the core Spring Framework, but support was added for async caching behavior none-the-less.

The _JCache_ library and overall API is an SPI (a specification, hence JSR-107), not a implementation or caching provider. This is not unlike Spring Framework's Cache Abstraction itself. As such, it is necessarily the lowest common denominator across an ecosystem of caching providers. By way of example, _Caffeine_, which provides async caching support, is also a JCache caching [provider implementation](https://github.com/ben-manes/caffeine/wiki/JCache).

Anyway, I happy to go in whatever direction we decide together.

jxblum commented on src/main/java/org/springframework/data/redis/cache/RedisCacheWriter.java:
```diff
@@ -135,6 +137,40 @@ default byte[] get(String name, byte[] key, @Nullable Duration ttl) {
+	 * @see java.nio.ByteBuffer
+	 * @since 3.2.0
+	 */
+	default Mono<byte[]> retrieve(String name, byte[] key) {
```
Right, I get it.

jxblum commented on src/main/java/org/springframework/data/redis/cache/RedisCache.java:
```diff
@@ -285,10 +290,31 @@ public void evict(Object key) {
 
+	@Override
+	public CompletableFuture<?> retrieve(Object key) {
+
```
Well, currently, true async is only supported if Lettuce is on the classpath and used as the Redis driver. If Jedis is used, then it requires help, as I designed and commented to above.

jxblum commented on src/main/java/org/springframework/data/redis/cache/RedisCacheWriter.java:
```diff
@@ -135,6 +137,40 @@ default byte[] get(String name, byte[] key, @Nullable Duration ttl) {
+	 * @return the {@link Mono value} to which the {@link RedisCache} maps the given {@link byte[] key}.
+	 * @throws IllegalStateException if the Redis connection factory is not reactive.
+	 * @see reactor.core.publisher.Mono
+	 * @see java.nio.ByteBuffer
```
Sorry, leftover documentation. Initially, I was using `ByteBuffers` for keys and then changed the `RedisCacheWriter.retrieve(..)` method signatures to align with the other `RedisCacheWriter` methods and operations (i.e. using a `byte[]` for keys).

My mistake.

jxblum commented:
@mp911de & @christophstrobl - I have completely reworked this PR and removed any strict dependency on Project Reactor. The API is simply tied to `CompletableFuture` now. 

However, I left the option to support Jedis in the mix.

As an alternative, I added 1 more [[OPTIONAL] commit](https://github.com/spring-projects/spring-data-redis/pull/2717/commits/5315fa7634e07899630280e29dcc724beae57cf0) demonstrating asynchronous caching behavior WITHOUT Jedis in the mix (i.e. only available when Lettuce is used), throwing the expected `UnsupportedOperationException` by delegating up to `Cache.retrieve(..)` if a non-reactive Redis driver is in use, such as Jedis.

I urge us to consider support for Jedis. But, if we decide not to support Jedis, then we simply need to merge the [OPTIONAL] commit and answer/remove the `TODOs` I placed in the codebase to make the questions/changes easier to find.

So, in summary, this PR is ready to merge once we decide which direction to go (with Jedis support or not).

jxblum commented on src/main/java/org/springframework/data/redis/cache/DefaultRedisCacheWriter.java:
```diff
@@ -59,6 +60,9 @@
 class DefaultRedisCacheWriter implements RedisCacheWriter {
 
+	private static final boolean REACTIVE_REDIS_CONNECTION_FACTORY_PRESENT = ClassUtils
+			.isPresent("org.springframework.data.redis.connection.ReactiveRedisConnectionFactory", null);
```
This is very ugly line-break. It read nicer as...

`ClassUtils.isPresent(..)`

on a single line.

jxblum commented on src/main/java/org/springframework/data/redis/cache/DefaultRedisCacheWriter.java:
```diff
@@ -419,40 +381,192 @@ private void checkAndPotentiallyWaitUntilUnlocked(String name, RedisConnection c
-			String message = String.format("Interrupted while waiting to unlock cache %s", name);
-
-			throw new PessimisticLockingFailureException(message, cause);
+			throw new PessimisticLockingFailureException(String.format("Interrupted while waiting to unlock cache %s", name),
```
Why do this? Now you have unnatural line break between the message the cause.

jxblum commented on src/main/java/org/springframework/data/redis/cache/RedisCache.java:
```diff
@@ -72,16 +71,16 @@ public class RedisCache extends AbstractValueAdaptingCache {
-	 * @param cacheConfiguration {@link RedisCacheConfiguration} applied to this {@link RedisCache} on creation;
-	 * must not be {@literal null}.
+	 * @param cacheWriter {@link RedisCacheWriter} used to perform {@link RedisCache} operations by executing the
+	 *          necessary Redis commands; must not be {@literal null}.
```
These Javadoc indentations are not consistent with the core Spring Framework Javadoc guidelines.

jxblum commented on src/main/java/org/springframework/data/redis/cache/RedisCache.java:
```diff
@@ -418,13 +429,23 @@ protected String convertKey(Object key) {
 
-		String message = String.format("Cannot convert cache key %s to String; Please register a suitable Converter"
-				+ " via 'RedisCacheConfiguration.configureKeyConverters(...)' or override '%s.toString()'",
+		String message = String.format(
```
Another unnatural line break.

jxblum commented on src/main/java/org/springframework/data/redis/cache/RedisCacheManager.java:
```diff
@@ -34,21 +34,20 @@
  * <p>
- * This {@link CacheManager} creates {@link Cache caches} on first write, by default. Empty {@link Cache caches}
- * are not visible in Redis due to how Redis represents empty data structures.
+ * This {@link CacheManager} creates {@link Cache caches} on first write, by default. Empty {@link Cache caches} are not
```
This is not appropriate places to create lines breaks in (source) Javadoc, regardless of how it is rendered.

I spent quite a bit of effort to clean this sort of thing up.

mp911de replied:
Javadoc formatting is automated, so we should not spend too much effort in manually aligning these as the formatter is going to apply its formatting anyways.

jxblum replied:
I don't agree with this statement. When I am working on, and maintaining source code, I am looking at the Javadoc in the source, checking for accuracy, necessary edits, completeness, etc. I am not looking to the rendered (e.g. HTML) form.

jxblum commented on src/test/java/org/springframework/data/redis/cache/RedisCacheTests.java:
```diff
@@ -15,18 +15,16 @@
-import static org.assertj.core.api.Assertions.assertThatIllegalStateException;
-import static org.assertj.core.api.Assumptions.assumeThat;
-import static org.awaitility.Awaitility.await;
+import static org.assertj.core.api.Assertions.*;
```
No star imports.

jxblum commented on src/test/java/org/springframework/data/redis/cache/RedisCacheTests.java:
```diff
@@ -353,8 +346,9 @@ void computePrefixCreatesCacheKeyCorrectly() {
 
-		doWithConnection(connection -> assertThat(connection.stringCommands()
-				.get("_cache_key-1".getBytes(StandardCharsets.UTF_8))).isEqualTo(binarySample));
+		doWithConnection(
```
Ugly line break.

jxblum commented on src/test/java/org/springframework/data/redis/cache/RedisCacheUnitTests.java:
```diff
@@ -46,76 +37,21 @@ void cacheRetrieveValueCallsCacheWriterRetrieveCorrectly() throws Exception {
 		verifyNoMoreInteractions(mockCacheWriter);
 	}
 
-	@Test // GH-2650
```
What happened to these test case methods?

mp911de replied:
These are testing internals (to some extent similar to testing getters and setters) and we have already tests for storing null values and null value caching in `RedisCacheTests` so it doesn't make sense to have redundancy causing additional maintenance efforts.

jxblum replied:
`RedisCacheTests` are "Integration" Tests whereas Unit Testing is a form of White-box testing. These tests are ensuring the correct handling of `null` values regardless of infrastructure concerns and are not equivalent to getter and setter testing.

I agree with removing redundancy.

jxblum commented:
These changes need further, closer review inside my IDE as they are far too extensive to review in GitHub alone.

jxblum commented:
This is now reviewed, polished and merged.
